Add Razor shims to ExternalAccess.Razor.

NTaylorMullen opened the pull request:
- This is the first step to removing IVT to Razor assemblies from core Roslyn assemblies. Once Razor is able to consume these changes and move to the new APIs we can entirely remove the IVTs from Roslyn assemblies.
- Brought Razor's `IDynamicFileInfoProvider` implementation over to the ExternalAccess.Razor package and built a new concept (`IRazorDocumentContainer`) that allows Razor to pass it just enough information about the document to Roslyn for it to "do the right thing".

dotnet/aspnetcore#19551

@tmat this takes care of all of the IVT dependent Razor APIs 🎁 

Razor PR to react to the changes (WIP): https://github.com/dotnet/aspnetcore-tooling/pull/1655

tmat commented on src/Tools/ExternalAccess/Razor/DocumentServiceProviderFactory.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
+{
+    internal abstract class DocumentServiceProviderFactory
```
>DocumentServiceProviderFactory [](start = 28, length = 30)

Can we call this `RazorDocumentServiceProviderFactory`? We generally prefix the ExternalAccess types with the name of the partner component, so that it's clear it's not a Roslyn concept.

tmat commented on src/Tools/ExternalAccess/Razor/DocumentServiceProvider.cs:
```diff
@@ -0,0 +1,72 @@
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
+{
+    internal class DocumentServiceProvider : IDocumentServiceProvider, IDocumentOperationService
```
>DocumentServiceProvider [](start = 19, length = 23)

Prefix with `Razor-`

tmat commented on src/Tools/ExternalAccess/Razor/DelegatingRazorDocumentExcerptService.cs:
```diff
@@ -0,0 +1,42 @@
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
+{
+    internal class DelegatingRazorDocumentExcerptService : IDocumentExcerptService
```
>DelegatingRazorDocumentExcerptService [](start = 19, length = 37)

Let's call this `RazorDelegatingDocumentExcerptService` for consistency.

tmat commented on src/Tools/ExternalAccess/Razor/DelegatingRazorSpanMappingService.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
+{
+    internal class DelegatingRazorSpanMappingService : ISpanMappingService
```
>DelegatingRazorSpanMappingService [](start = 19, length = 33)

 `RazorDelegatingSpanMappingService`

tmat commented on src/Tools/ExternalAccess/Razor/DefaultDocumentServiceProviderFactory.cs:
```diff
@@ -0,0 +1,30 @@
+{
+    [Shared]
+    [Export(typeof(DocumentServiceProviderFactory))]
+    internal class DefaultDocumentServiceProviderFactory : DocumentServiceProviderFactory
```
>DefaultDocumentServiceProviderFactory [](start = 19, length = 37)

Is this needed? I don't see any imports in the ASP.NET PR.

tmat commented on src/Tools/ExternalAccess/Razor/DefaultDocumentServiceProviderFactory.cs:
```diff
@@ -0,0 +1,30 @@
+{
+    [Shared]
+    [Export(typeof(DocumentServiceProviderFactory))]
+    internal class DefaultDocumentServiceProviderFactory : DocumentServiceProviderFactory
```
>internal class DefaultDocumentServicePr [](start = 4, length = 39)

Couldn't this be kept in ASP.NET repo? It's good to minimize the amount of code here.

NTaylorMullen replied:
Eh, we need pieces of this here. I've put the core guts of this in the ASP.NET Core repo but I had to leave one of these here in order to allow Roslyn to request things like `ISpanMappingService` and then for this to request `IRazorSpanMappingService` from an inner provider and wrap the result in a mapping service wrapper

tmat commented on src/Tools/ExternalAccess/Razor/DelegatingRazorSpanMappingService.cs:
```diff
@@ -0,0 +1,37 @@
+                roslynSpans[i] = new MappedSpanResult(razorSpan.FilePath, razorSpan.LinePositionSpan, razorSpan.Span);
+            }
+
+            return roslynSpans.ToImmutableArray();
```
>roslynSpans [](start = 19, length = 11)

Can be simplified to `razorSpans.SelectAsArray(span => new MappedSpanResult(span.FilePath, span.LinePositionSpan, span.Span));` using [`SelectAsArray`](http://sourceroslyn.io/#Microsoft.CodeAnalysis/Collections/ArrayBuilderExtensions.cs,71) extension method.

NTaylorMullen replied:
Eh, I like to avoid the terser statements like that because it's "less" code but harder to visibly consume/debug through.

tmat commented on src/Tools/ExternalAccess/Razor/IRazorIDocumentExcerptService.cs:
```diff
@@ -0,0 +1,47 @@
+        Tooltip
+    }
+
+    internal struct RazorExcerptResult
```
>struct [](start = 13, length = 6)

`readonly`

NTaylorMullen replied:
I'll apply the wrapper pattern to other pieces but this specific type is what Razor needs to new-up and return.

tmat commented on src/Tools/ExternalAccess/Razor/FilePathComparer.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
+{
+    internal static class FilePathComparer
```
>FilePathComparer [](start = 26, length = 16)

Perhaps this could be nested in RazorDynamicFileInfoProvider?

tmat commented on src/Tools/ExternalAccess/Razor/FilePathComparer.cs:
```diff
@@ -0,0 +1,31 @@
+                    _instance = StringComparer.OrdinalIgnoreCase;
+                }
+
+                return _instance;
```
Move this as a method to RazorDynamicFileInfoProvider?

tmat commented on src/Tools/ExternalAccess/Razor/EmptyTextLoader.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
+{
+    internal class EmptyTextLoader : TextLoader
```
>EmptyTextLoader [](start = 19, length = 15)

Perhaps this could be nested in RazorDynamicFileInfoProvider?

tmat commented on src/Tools/ExternalAccess/Razor/EmptyTextLoader.cs:
```diff
@@ -0,0 +1,30 @@
+        public EmptyTextLoader(string filePath)
+        {
+            _filePath = filePath;
+            _version = VersionStamp.Create(); // Version will never change so this can be reused.
```
>Create() [](start = 36, length = 8)

Isn't `VersionStamp.Default` more appropriate?

NTaylorMullen replied:
Could be, just brought over from Razor land 😄

NTaylorMullen commented:
🆙 📅 

Refactored Razor.ExternalAccess to leave more logic in Razor.

- Now `IDynamicFileInfoProvider` is shimmed and the core-logic exists in the Razor repo.
- Added the ExternalAccess.Razor piece to the Roslyn deployment to ensure our `IDynamicFileInfoProvider` wrapper can be found by MEF components. In the case that the Razor workload is not available we allow a `null` Razor dynamic file implementation.
- Updated the ExternalAccess.Razor project to be netstandard2.0 instead of net472 so it vibes better with the Razor repo.

tmat commented on src/Tools/ExternalAccess/Razor/RazorDynamicFileInfoProviderWrapper.cs:
```diff
@@ -0,0 +1,67 @@
+
+        [ImportingConstructor]
+        public RazorDynamicFileInfoProviderWrapper(
+            [Import(AllowDefault = true)] IRazorDynamicFileInfoProvider innerDynamiFileInfoProvider)
```
>innerDynamiFileInfoProvider [](start = 72, length = 27)

typo - missing `c`.

tmat replied:
We decided that we'll keep using MEF component rejection, so this doesn't need to be `AllowDefault = true`. It should however be `Lazy<IRazorDynamicFileInfoProvider>`, so that we don't load Razor in projects that are not Razor projects.

---
In reply to: [392333118](https://github.com/dotnet/roslyn/pull/42174#discussion_r392333118) [](ancestors = 392333118)

NTaylorMullen commented:
🆙 📅 

Turn nullable reference types on for ExternalAccess.Razor and make the DynamicFileInfoProvider not load Razor assemblies in non-Razor scenarios.
